core/src/main/java/org/apache/struts2/interceptor/csp/DefaultCspSettings.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.struts2.interceptor.csp;
+
+import com.opensymphony.xwork2.ActionContext;
+
+import javax.servlet.http.HttpServletResponse;
+import java.security.SecureRandom;
+import java.util.Base64;
+import java.util.Map;
+
+
+public class DefaultCspSettings implements CspSettings {
+
+    private static final int NONCE_LENGTH = 18;
+    private final SecureRandom sRand  = new SecureRandom();
+    private String reportUri = "cspviolation";
+    private boolean reporting = true;
+    //TODO add string constants for csp header to avoid doing string operations each time
+
+    public void setReporting(boolean value){
+        this.reporting = value;
+    }
+
+    public void addCspHeaders(HttpServletResponse response) {
+        createNonce();
+        if (this.reporting){
+            response.setHeader(CSP_REPORT_HEADER, getPolicyString());
+        } else {
+            response.setHeader(CSP_ENFORCE_HEADER, getPolicyString());
+        }
+    }
+
+    public void setReportUri(String reportUri) {
+        this.reportUri = reportUri;
+    }
+
+    public String getNonceString() {
+        Map<String, Object> session = ActionContext.getContext().getSession();
+        return (String) session.get("nonce");
+    }
+
+    protected void createNonce() {
+        String nonceValue = Base64.getUrlEncoder().encodeToString(getRandomBytes(NONCE_LENGTH));
+        Map<String, Object> session = ActionContext.getContext().getSession();
+        session.put("nonce", nonceValue);
+    }
+
+    private String getPolicyString() {
+        return String.format("%s '%s'; %s 'nonce-%s' '%s' %s %s; %s '%s'; %s '%s';",
+                OBJECT_SRC, NONE,
+                SCRIPT_SRC, getNonceString(), STRICT_DYNAMIC, HTTP, HTTPS,
```

Comment:
super nit: Just as a sanity check, both getNonceString() and this.reportUri won't have ' as a character that might potentially mess with the quoting in the template string, right? Is there any way for the user to affect these values, either maliciously or by mistake?

Reply:
[According to the spec](https://www.w3.org/TR/CSP3/#framework-directive-source-list) the final nonce should be a base64 encoded version of the random string. Base64's character set is orthogonal to that of structure headers!

Reply:
We found this error for this.reporrtUri but good catch for the nonce too!